### PR TITLE
Lib cache

### PR DESCRIPTION
**Pull request by rdaly525**
Needed for:
https://github.com/rdaly525/coreir/pull/930

### CODE REVIEW

File: coreir/context.py [hunk 1]
```diff
@@ -51,6 +52,7 @@ def __init__(self, ptr=None):
         if ptr is None:
             self.external_ptr = False
             ptr = libcoreir_c.CORENewContext()
+            _LIBRARY_CACHE.clear()
```
Comment (leonardt) [inline coreir/context.py:55]: Should we make the library cache per-context object? Not that we ever do this, but wouldn't this break if we had two separate context objects (with two separate CoreIR contexts)? I think one tricky thing is that we could multiple pycorier contexts for the same coreir contexts, so maybe we could make _LIBRARY_CACHE a map from ptr -> cache (so each coreir context ptr object has a unique cache

File: coreir/context.py [hunk 2]
```diff
@@ -219,6 +225,7 @@ def compile_to_verilog(self, top, filename, libs=(), split="",
 
     def __del__(self):
         if not self.external_ptr:
+            print("!!!Deleting Context!!!")
```
Comment (leonardt) [inline coreir/context.py:228]: Should we leave this print in? We could switch it to use the logging module for debugging purposes

File: coreir/context.py (hunk 1 above)
Comment (rdaly525) [reply coreir/context.py:55]: I initially had that, but not sure how to cache the ptr (it is unhashable) and did not think this situation would occur.

Comment (leonardt) [reply coreir/context.py:55]: Maybe we can use ctypes.addressof to get the address of the underlying pointer and use that to hash? I think that should be safe.

Here's an example: https://github.com/leonardt/pycoreir/blob/5601216821931cfd036c05f2827992b9c3b7dce0/coreir/type_gen.py#L19

Comment (leonardt) [reply coreir/context.py:55]: I think here you can try using `ctypes.addressof(ptr.contents)`

Comment (rdaly525) [reply coreir/context.py]: Fixed

File: coreir/context.py [hunk 3]
```diff
@@ -52,7 +52,8 @@ def __init__(self, ptr=None):
         if ptr is None:
             self.external_ptr = False
             ptr = libcoreir_c.CORENewContext()
-            _LIBRARY_CACHE.clear()
+            _LIBRARY_CACHE.setdefault(ct.addressof(ptr), {})
+        assert not self.external_ptr
```
Comment (leonardt) [inline coreir/context.py:56]: Is this assertion needed?

Comment (rdaly525) [reply coreir/context.py]: fixed

**Pull request merged**